### ISSUE TITLE
Friendship 里增加 申请好友的”来源“

### ISSUE THREAD

**Issue by nickwong** [Chatie/wechaty#1801]
**Is your feature request related to a problem? Please describe.**
处理好友申请时，通过Friendship.hello()可以获取到验证信息，但Friendship里无法获取到用户”来源“。
希望能够加上”来源“，这样在用户通过名片添加机器人的时候，可以通过”来源“知道是哪个用户推介的。

**Describe the solution you'd like**
在Friendship里增加获取来源的方法。

**Describe alternatives you've considered**

**Additional context**

[enhancement]

**Comment by lijiarui** [comment]
@nickwong PR is welcome!

**Comment by quhuainan** [comment]
强烈希望能添加这个功能。真的好重要

**Comment by quhuainan** [comment]
悬赏300块红包，开发这个功能。做完了，可以找我领。微信：15510974083

**Comment by su-chang** [comment]
对于该功能的支持，给大家提供一些参考。

- 目标

  需要给 wechaty 新增一个接口来获取收到好友请求的好友来源。

- 数据依赖

  > 具体数据不贴在这里，持有 padplus token 的小伙伴可以通过以下命令启动程序并查看详细数据
  ```js
  BROLOG_LEVEL=silly ts-node index.ts
  ```
  自动通过好友后，会接收到底层发来的一条消息，通过其中的 `RecommendInfo` 字段可以获取好友请求的详细信息，如：微信号，昵称，来源等。

- 数据解释

  `RecommendInfo`该字段所对应的对象中 `scene` 字段表示好友的来源，具体对应关系如下：

  1：QQ
  2：邮箱
  3：微信号
  14：群聊
  15：手机号
  18：附近的人
  25：漂流瓶
  29：摇一摇
  30：二维码

欢迎感兴趣的小伙伴进行开发，有任何相关问题均可在该 issue 中进行交流讨论。

**Comment by H3lloTom** [comment]
看了一下，发现是wechaty-puppet内定义了friendshipPayload方法，puppet-padplus内实现了该方法，是否要对wechaty-puppet和其他的协议实现进行同步修改 @su-chang 
[wechaty-puppet](https://github.com/wechaty/wechaty-puppet/blob/master/src/schemas/friendship.ts#L53)
[wechaty-puppet-padplus](https://github.com/wechaty/wechaty-puppet-padplus/blob/master/src/pure-function-helpers/friendship-raw-payload-parser.ts#L107)

**Comment by su-chang** [comment]
@H3lloTom you are right.

**Comment by H3lloTom** [comment]
我想提交这样的改动，在wechaty-puppet的schema上加一个可选属性?sharedContactId，然后在wechaty-puppet-padplus的friendShipParser内将该属性的值加上。 @su-chang

**Comment by su-chang** [comment]
@H3lloTom You could have a draft design, and send a PR to these repos, thank you very much.

**Comment by H3lloTom** [comment]
i have created a pr ,please take a review  ^-^  @huan @su-chang 
https://github.com/wechaty/wechaty-puppet/pull/102

**Comment by huan** [comment]
@H3lloTom  Hi! Thank you very much for the PR, it seems that we do need to record the source of the friendship in our puppet payload!

I'm sorry for the late reply because we are busy working on the multi-language Wechaty recently. I'll try my best to go back to this PR later after we published all the three new language wechaty.

Thank you again for your patient, I hope we can add this enhancement soon!

**Comment by H3lloTom** [comment]
> @H3lloTom Hi! Thank you very much for the PR, it seems that we do need to record the source of the friendship in our puppet payload!
> 
> I'm sorry for the late reply because we are busy working on the multi-language Wechaty recently. I'll try my best to go back to this PR later after we published all the three new language wechaty.
> 
> Thank you again for your patient, I hope we can add this enhancement soon!

haha, it's ok!

**Comment by huan** [comment]
@H3lloTom Hello, I'm reviewing your PR about adding `sharedContactId` in friendship payload now, I have a question for this PR:

We have already a `scene` property in the friendship payload, and according to the comment from @su-chang , we have an enum for the source of friendship too.

So why do we need a `sharedContactId` in our friendship payload? What will it store and what it for?

Please explain so that I can understand what you need before we can continue discussing this PR.

**Comment by Seven-Y** [comment]
hello。该问题是否得到解决。后续JAVA等版本是否将同步跟进。

查看了相应的API。并未涉及相关的功能

**Comment by suhli** [comment]
@huan 
we need information of this
![image](https://user-images.githubusercontent.com/23626907/104830399-c3141780-58b9-11eb-8148-98d5a8f13300.png)

**Comment by huan** [comment]
@suhli Thank you very much for your Pull Requests for adding this great feature to wechaty, wechaty-puppet, and wechaty-puppet-padlocal !

I have a question: as your screenshot shows that the **source** might be from a contact card sharing, can we enum all potential _sources_ when we add a new contact?

Now we have 1, what are the other types?

**Comment by suhli** [comment]
@huan i can't find another scene that contains those keys,but i found a scene that not emued at FriendshipSceneType with value 48(radar) 
![image](https://user-images.githubusercontent.com/23626907/104935602-66fae180-59e6-11eb-858f-c0091984cdeb.png)

**Comment by huan** [comment]
Yes, the `Radar` scene is the source type that I believe should be put into the Source Enum.

Could you please post the JSON of the Object which you screenshotted as text in the issue? Because that would be more readable.

I did a quick check for the FriendShip Source in my WeChat app, I believe it should at least contains the following scenes:

1. ContactCard: from XXX contact
2. Radar
3. Search: WeiXin ID
4. Search: Mobile Phone Number
5. Search: QQ
4. Room: in XXX room
5. please add...

And I'm just wondering if there's any place that we can read them from the code...

**Comment by suhli** [comment]
> Yes, the `Radar` scene is the source type that I believe should be put into the Source Enum.
> 
> Could you please post the JSON of the Object which you screenshotted as text in the issue? Because that would be more readable.
> 
> I did a quick check for the FriendShip Source in my WeChat app, I believe it should at least contains the following scenes:
> 
> 1. ContactCard: from XXX contact
> 2. Radar
> 3. Search: WeiXin ID
> 4. Search: Mobile Phone Number
> 5. Search: QQ
> 6. Room: in XXX room
> 7. please add...
> 
> And I'm just wondering if there's any place that we can read them from the code...
> 
> BTW: I'm also thinking about the name conversion about the FriendShip Source: does it makes more sense if rename to `Scene`? Does it math the existing Scene enum?

got it

**Comment by skys215** [comment]
Any progress on this request?

**Comment by suhli** [comment]
> Any progress on this request?

seem's no
